codegen/smithy-aws-swift-codegen/src/test/kotlin/software/amazon/smithy/aws/swift/codegen/customizations/RulesBasedAuthSchemeResolverGeneratorTests.kt

```diff
@@ -0,0 +1,151 @@
+                            var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4")
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: "weather")
+                            guard let region = serviceParams.region else {
+                                throw ClientError.authError("Missing region in auth scheme parameters for SigV4 auth scheme.")
+                            }
+                            sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: region)
+                            validAuthOptions.append(sigV4Option)
+                    }
+                    return validAuthOptions
+                }
+
+                public func constructParameters(context: HttpContext) throws -> ClientRuntime.AuthSchemeResolverParameters {
+                    return try DefaultS3AuthSchemeResolver().constructParameters(context: context)
+                }
+            }
+
+            public struct DefaultS3AuthSchemeResolver: S3AuthSchemeResolver {
+                public func resolveAuthScheme(params: ClientRuntime.AuthSchemeResolverParameters) throws -> [AuthOption] {
+                    var validAuthOptions = Array<AuthOption>()
+                    guard let serviceParams = params as? S3AuthSchemeResolverParameters else {
+                        throw ClientError.authError("Service specific auth scheme parameters type must be passed to auth scheme resolver.")
+                    }
+                    let endpointParams = EndpointParams(authSchemeParams: serviceParams)
+                    let endpoint = try DefaultEndpointResolver().resolve(params: endpointParams)
+                    guard let authSchemes = endpoint.authSchemes() else {
+                        return try InternalModeledS3AuthSchemeResolver().resolveAuthScheme(params: params)
+                    }
+                    let schemes = try authSchemes.map { (input) -> AWSClientRuntime.AuthScheme in try AWSClientRuntime.AuthScheme(from: input) }
+                    for scheme in schemes {
+                        switch scheme {
+                            case .sigV4(let param):
+                                var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4")
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: param.signingName)
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: param.signingRegion)
+                                validAuthOptions.append(sigV4Option)
+                            case .sigV4A(let param):
+                                var sigV4Option = AuthOption(schemeID: "aws.auth#sigv4a")
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingName, value: param.signingName)
+                                sigV4Option.signingProperties.set(key: AttributeKeys.signingRegion, value: param.signingRegionSet?[0])
```

Comment:
is signingRegion required? what will happen if param.signingRegionSet is null or first item is null

Reply:
If `param.signingRegionSet` is null or first item is null, then that would mean the service team failed to model their endpoint ruleset correctly. Signing region is required for SigV4.

Reply:
in that case should we catch if its null and throw an error accordingly? right now it would be passed in even if null